app/models/task.rb

```
@@ -170,6 +170,49 @@ def self.modify_params(params)
     params
   end
 
+  def update_task_type(params)
+    ActiveRecord::Base.transaction do
+      new_branch_task = first_ancestor_of_type.change_type(params)
+      new_child_task = new_branch_task.last_descendant_of_type
+      new_child_task.update!(status: status)
+
+      if assigned_to.is_a?(User) && new_child_task.assigned_to.is_a?(User) &&
+         parent.assigned_to_same_org?(new_child_task.parent)
+        new_child_task.update!(assigned_to: assigned_to)
+      end
+
+      children.active.each { |child| child.update!(parent_id: last_descendant_of_type.id) }
+      first_ancestor_of_type.cancel_descendants
+
+      [first_ancestor_of_type.descendants, new_branch_task.first_ancestor_of_type.descendants].flatten
+    end
+  end
+
+  def assigned_to_same_org?(task_to_check)
+    assigned_to.is_a?(Organization) && assigned_to == task_to_check.assigned_to
+  end
+
+  def first_ancestor_of_type
+    (parent && same_task_type?(parent)) ? parent.first_ancestor_of_type : self
+  end
+
+  def last_descendant_of_type
+    child_of_task_type = children.open.detect { |child| same_task_type?(child) }
```

Can we raise an error if there are multiple children for the current task? I think the behaviour in that case is a little unexpected (that it would select the first--by ID? created_at timestamp?--open child task).